src/abstracts/SablierV2AirstreamCampaign.sol

```diff
@@ -0,0 +1,127 @@
+
+/// @title SablierV2AirstreamCampaign
+/// @notice See the documentation in {ISablierV2AirstreamCampaign}.
+abstract contract SablierV2AirstreamCampaign is
+    ISablierV2AirstreamCampaign, // 2 inherited component
+    Adminable // 1 inherited component
+{
+    using BitMaps for BitMaps.BitMap;
+    using SafeERC20 for IERC20;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                               USER-FACING CONSTANTS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ISablierV2AirstreamCampaign
+    IERC20 public immutable override asset;
+
+    /// @inheritdoc ISablierV2AirstreamCampaign
+    bool public immutable override cancelable;
+
+    /// @inheritdoc ISablierV2AirstreamCampaign
+    uint40 public immutable override expiration;
+
+    /// @inheritdoc ISablierV2AirstreamCampaign
+    bytes32 public immutable override merkleRoot;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                  INTERNAL STORAGE
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @dev Packed booleans that record the history of claims.
+    BitMaps.BitMap private _claimedBitMap;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                    CONSTRUCTOR
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @dev Constructs the contract by initializing the immutable state variables.
+    constructor(address initialAdmin, IERC20 asset_, bytes32 merkleRoot_, uint40 expiration_, bool cancelable_) {
+        admin = initialAdmin;
+        asset = asset_;
+        merkleRoot = merkleRoot_;
+        expiration = expiration_;
+        cancelable = cancelable_;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                           USER-FACING CONSTANT FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ISablierV2AirstreamCampaign
+    function hasClaimed(uint256 index) public view override returns (bool) {
+        return _claimedBitMap.get(index);
+    }
+
+    /// @inheritdoc ISablierV2AirstreamCampaign
+    function hasExpired() public view override returns (bool) {
+        return expiration > 0 && expiration <= block.timestamp;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                         USER-FACING NON-CONSTANT FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ISablierV2AirstreamCampaign
+    function clawback(address to, uint128 amount) external override onlyAdmin {
+        // Checks: the campaign has expired.
+        if (!hasExpired()) {
+            revert Errors.SablierV2AirstreamCampaign_CampaignExpired({
```

Comment:
The name of this error should be `CampaignNotExpired`.

Reply:
Good catch